Alex/cleanup add tests etc

alex-harvey-z3q opened the pull request.

alex-harvey-z3q committed:
Uplift kms.tf to use 0.12 first class expressions

alex-harvey-z3q committed:
Add boilerplate and code generated docs

Adds missing boilerplate files, files for code generated docs, and
completes the README up to standard.

alex-harvey-z3q committed:
Bugfix in variables2md.rb

alex-harvey-z3q committed:
Add examples and tests

alex-harvey-z3q committed:
Create role dependency in test

Without this patch applied, terraform apply would fail with:

  Error: MalformedPolicyDocumentException: Policy contains a statement with one or more invalid principals.
        status code: 400, request id: 966acdeb-2ef8-4329-be20-f0d64293d181

The issue being that the IAM role that this module depends upon was not
present.

alex-harvey-z3q committed:
Bugfix in kms.tf

Without this patch applied, terraform apply would fail with:

  Error: MalformedPolicyDocumentException: The new key policy will not allow you to update the key policy in the future.
        status code: 400, request id: 39345ff4-12d5-48db-9197-21b849dac8be

On further investigation, I found that logic in the locals section was
making assumptions about the caller identity.

In my case, the caller identity was:

  arn:aws:iam::111111111111:user/alex

On further investigation, it appeared there was no need for the logic
around this anyway, thus this patch.

alex-harvey-z3q committed:
Allow root to administer the key

Without this patch applied, the root user cannot manage the KMS key,
which doesn't appear to be the intent.

alex-harvey-z3q committed:
Be consistent with commas in arrays

alex-harvey-z3q committed:
Add a test

alex-harvey-z3q committed:
Add some outputs

alex-harvey-z3q committed:
Minor bugfix in examples/shunit2/test_helper.sh

alex-harvey-z3q commented on erb/docs.rb:
```diff
@@ -0,0 +1,4 @@
+require 'erb'
+template = File.read('erb/README.erb')
+renderer = ERB.new(template, nil, '-')
+File.write('README.md', renderer.result())
```
@aarongorka in case you're wondering, all this Ruby supports only the code generation in the docs and it would be moved into Cookiecutter once I figure out where that is and how it works. Also, in this case, this code really does need to be written in Ruby- it's just not possible to do what I'm doing here using Python/Jinja2 - I [tried](https://alexharv074.github.io/2018/11/09/generating-markdown-from-source-code-in-erb.html).

alex-harvey-z3q commented on examples/shunit2/test_helper.sh:
```diff
@@ -0,0 +1,42 @@
+
+vars=(
+  AWS_ACCESS_KEY_ID
+  AWS_SECRET_ACCESS_KEY
```
@aarongorka , please note that use of access keys in the tests is only temporary until I figure out a better way based on using Roles etc.

aarongorka commented:
> @aarongorka are you ok with this?

Hey @alexharv074, it doesn't work because you're not assuming a role. When you do, the datasource/API call returns a `assumed-role` ARN (with the session name in it too), but KMS needs to have the normal role ARN like I had constructed with the locals. We don't really use IAM users, but you can include it anyway (maybe leave a comment stating why).

alex-harvey-z3q committed:
Restore original kms.tf but with 0.12 syntax

alex-harvey-z3q committed:
Add fix to support non role based access

In commit 3cc7231a36bc990cc76793a37a74c59407371661, an incorrect
assumption was applied. In this correction, I use the regexall feature
of Terraform 0.12.7 to detect a "/" in the current_account_arn.

If not there, I assume the user has not used a role-based authentication.

alex-harvey-z3q commented on kms.tf:
```diff
@@ -1,20 +1,20 @@
-  my_role_arn  = "arn:aws:iam::${data.aws_caller_identity.current.account_id}:role/${local.my_role_name}"
+  current_arn        = data.aws_caller_identity.current.arn
+  current_account_id = data.aws_caller_identity.current.account_id
+  my_role_arn        = length(regexall("/", local.current_arn)) > 0 ? "arn:aws:iam::${local.current_account_id}:role/${split("/", local.current_arn)[1]}" : local.current_arn
```
@aarongorka , I finally got back to this. Do you think this will work? It introduces a dependency on 0.12.7 which I think is ok, and that length/regexall is Terraform craziness for, "does this string contain regex "/"" ? So basically, if the current_arn contains a forward slash, use your expression ; otherwise, just use the current_arn and assume we logged in by keys etc.
